New_EU_presets

NomDeTom opened the pull request:
Some new presets for a wider range of operations in the EU. Ham bands are now removed, pending further discussion.
Relies on https://github.com/meshtastic/protobufs/pull/824 for protobufs.

## What has changed?
- 2 new region presets
  - EU_866, 
  - NARROW_868,
- 4 new modem presets
  - LITE_FAST/SLOW, 
  - NARROW_FAST/SLOW
- new region parameters
  - licensedOnly, to limit a region to ham operators, i.e. those operating with the licensed flag active - placeholder for now.
  - overrideSlots, to allow a region to always default to a particular slot
  - defaultPreset, to allow a default other than LongFast
  - availablePreset, to allow limiting preset availability to a particular region
- refactoring of the modem setting checks to happen before restarting the node
- adding additional client notifications for invalid setting combinations

## What is needed to integrate these changes more widely
- BaseUI / InkHud
  - The new regions and presets have been added to the menu lists, but the list logic has not been updated to exclude invalid combinations
- MUI
  - The new regions and presets need to be added to the menu lists, and the list logic included.
- Client Apps
  - These will need updating to take account of the valid region/preset combinations, ideally hiding invalid combinations within the menus
  - Their display logic will need updating to include the new overrideSlots information, which will affect both the lora config menu and the behaviour when new primary channels are generated

## Acknowledgements
This change has been supported extensively by @Stary2001, @phaseloop, and @caveman99. Their input cannot be understated.

## 🤝 Attestations

- [x] I have tested that my proposed changes behave as described.
- [ ] I have tested that my proposed changes do not cause any obvious regressions on the following devices:
  - [ ] Heltec (Lora32) V3
  - [ ] LilyGo T-Deck
  - [ ] LilyGo T-Beam
  - [ ] RAK WisBlock 4631
  - [ ] Seeed Studio T-1000E tracker card
  - [x] Other (please specify below)
Pro-micro DIY

caveman99 committed:
Add a new frequency range for EU SRD area with adaptive DC

NomDeTom committed:
new presets yo

NomDeTom committed:
Update protobufs

NomDeTom committed:
Merge pull request #4 from NomDeTom/create-pull-request/update-protobufs

Update protobufs and classes

NomDeTom committed:
guards against using presets outside of region

NomDeTom commented on src/mesh/RadioInterface.cpp:
```diff
@@ -518,6 +553,31 @@ void RadioInterface::applyModemConfig()
+                cr = 6;
+                sf = 8;
+                break;
+            case meshtastic_Config_LoRaConfig_ModemPreset_HAM_FAST:
```
The intention here is to give the HAM_US433 a specific set of presets for use in that band, which consists of the two narrow settings (already in use) plus an equivalent of shortfast for use at meetups. I should emphasise that this particular setup has been handed to me, so I'm not familiar with the details.

NomDeTom replied:
@wehooper4 what was the intended setting for the "wide" ham preset?

NomDeTom committed:
update EU 866MHz band description

Stary2001 committed:
Refactor preset validation to happen before applying radio config (and don't turn off Bluetooth if it's invalid)

NomDeTom committed:
add default preset to regions

NomDeTom committed:
update RDEF macro to include default presets and preset bits

NomDeTom committed:
Refactor RDEF macro for improved readability and consistency; adjust spacing in modem configuration functions.

NomDeTom committed:
Improved the allowed presets checking by adding a bitfield to each region. Corrected some text. Inserted todo items.

NomDeTom commented on src/mesh/MeshRadio.h:
```diff
@@ -5,6 +5,32 @@
+    bool allowPreset_HAM_FAST;    // 500kHz BW
+    bool reserved;
+};
+
```
Thanks - this is the kind of insight I needed.

I'll make the changes later.

Do you have any view on how to iterate the presets list for the menu display?

Stary2001 committed:
Check settings against the new region, not the old one

phaseloop committed:
use array of accepted modes per region

phaseloop committed:
fix compliation error

NomDeTom committed:
Merge pull request #6 from phaseloop/New_presets-array

use array of accepted modes per region

NomDeTom committed:
added ham check

phaseloop commented on src/mesh/RadioInterface.cpp:
```diff
@@ -279,10 +267,15 @@ void initRegion()
     for (; r->code != meshtastic_Config_LoRaConfig_RegionCode_UNSET && r->code != code; r++)
         ;
+    if (r->licensedOnly == true && !devicestate.owner.is_licensed) {
+        LOG_WARN("Region code %d not permitted without license, defaulting to unset", config.lora.region);
```
This should be checked in validateModemConfig and return to default band preset as we do for any other problem with presets. Otherwise someone will end with bricked device with no lora if they try to set it remotely

NomDeTom committed:
interim for review

wehooper4 committed:
Update HAM US433 frequency range and config settings

Updated US ham mode to full legal frequency range

Set hamfast to a legal setting

NomDeTom committed:
Merge pull request #7 from wehooper4/patch-1

Update HAM US433 frequency range and config settings

NomDeTom committed:
remove hamfast

NomDeTom committed:
Update protobufs

NomDeTom committed:
Merge pull request #8 from NomDeTom/create-pull-request/update-protobufs

Update protobufs and classes

NomDeTom committed:
tidy up removing hamfast

NomDeTom committed:
fixed license check

NomDeTom committed:
Add overrideSlot to RegionInfo and update RDEF macro usage, added error messages to client notification

NomDeTom committed:
added logging if slot override is being used

NomDeTom committed:
Merge upstream branch changes, inc KZ region update and custom CR changes

NomDeTom committed:
trunk FML

NomDeTom commented:
@jp-bennett This PR now includes your custom CR check. Can you please re-check it in the context of this change? I don't know if we need additional logic to handle if moving to a preset with a higher default preset than the custom CR set, or indeed if we need to take account of people wanting to run CR4.

In any case, can you help me avoid a reversion of your change?

NomDeTom committed:
fixed error in CR check, added logic for if a lower CR than default is used.

NomDeTom committed:
Merge branch 'develop' into New_presets

NomDeTom committed:
Remove ham settings for now.

NomDeTom committed:
Update protobufs

NomDeTom committed:
Merge pull request #10 from NomDeTom/create-pull-request/update-protobufs

Update protobufs and classes

NomDeTom committed:
Update protobuf submodule to latest commit

NomDeTom committed:
Update protobufs

NomDeTom committed:
Merge pull request #11 from NomDeTom/create-pull-request/update-protobufs

Update protobufs and classes

NomDeTom committed:
Update protobuf submodule to latest commit

NomDeTom committed:
Update protobufs

NomDeTom committed:
Merge pull request #12 from NomDeTom/create-pull-request/update-protobufs

Update protobufs and classes

NomDeTom committed:
Merge branch 'develop' into New_presets

NomDeTom committed:
Fix preset configuration for EU_866 region and update channel details for NARROW_868

NomDeTom committed:
Enhance region definitions by adding text, position, and telemetry throttle parameters

NomDeTom committed:
tunk

NomDeTom committed:
Revert protobuf change

NomDeTom committed:
Merge remote-tracking branch 'origin/develop' into New_presets

NomDeTom commented:
@psimolin @j0uni @tekomula @oherrala @petrisi in case you're not subscribed here, too.

NomDeTom committed:
Merge remote-tracking branch 'origin/develop' into New_presets

rob-fi commented:
> I won't respond to all the comments and questions here and in #8855 and #7481, but I will explain as much as I can.

Thanks so much for writing that @NomDeTom. Although tedious to write out, it's really helpful to understand the logic and reasoning for implmenenting things in certain ways. There's some great information there to digest, and I/perhaps others will do a bit of digging (in particular with the spectral density charactaristics of LoRa modulation) and see if we come to the same conclusions, but your reasoning seems sound enough based on what's written.

Although there are many SRD devices on-band that, let's just say, aren't too strict with compliance, there's always the possibility to lead by example, especially with a project of this nature that now has become so popular. We don't want to make excuses for it being outlawed. At the same time, if there's a way to legally cater for a not insignificant user base already using particular settings, then it's worth exploring a little.

NomDeTom committed:
padding added to supplement spacing of channels

NomDeTom committed:
added more commentary

NomDeTom committed:
merge conflicts fixed - TY Stary

rob-fi commented:
Thanks for the insight @derpyspike!

But yes the collision with other LoRa devices is a concern because so many are generally operating in the middle of the band, so keeping well away from them is an important consideration. I believe that was the whole reason the EdgeFastLow "setting" came about. There could be a 3rd channel in the middle of course if spacing was really wished, but having 4 makes more sense.

I just hope that this topic can remain open and with good discussion going forward to determine what is absoutely possible or not. It's my worry that this just gets barged through and merged without any further consideration, so I'll try to flag this for discussion as much as I can both here and in Discord, and I'll try to consult with some experts on the subject too. There's discussions brewing about a Meshtastic fork that would include such edge of frequency settings and that's the last thing I want to see happening.

With regards to the meshtastic collision of adjacent channels, well, having devices next to each other is going to do that :D. That's unfortunately a general RF problem for more than just meshtastic!

By the way here's an illustration of a 62.5kHz bandwidth meshtastic message with the marker set to 62.5kHz bandwith also. This a completely uncalibrated (other than the ppm offset of the receiver adjusted) measurement of a node a few meters away outside a window (in -15°C, clearly has TCXO), so the signal is very strong. But it serves to illustrate how the energy is concentrated within the bandwidth.

<img width="815" height="366" alt="efl" src="https://github.com/user-attachments/assets/1295cdca-8ef3-44eb-9ac7-8e7beefe8ab7" />

rob-fi commented:
> > However, what I will say is that it questionable about enforcing the limits for these specific presets if the existing presets also turn out to be in violation of the regulations, and there's no desire to immediately correct that.
> 
> I have little to no technical knowledge to contribute here, but want to point out that what you're saying is that; because the old presets might not be compliant, no effort should be made to make new presents comply either. This is in general a pretty dangerous stance.

The complete opposite :)

I was saying that there should be an immediate desire to correct them also, as well as informing those using "EFL" and also Meshcore that they're breaching regulations, if it turns out to be the case. I'm only trying to establish with absolute certainty what's in compliance and what isn't.

NomDeTom committed:
Add MeshRadio includes and region-specific throttling checks in telemetry modules

NomDeTom commented on src/mesh/MeshRadio.h:
```diff
@@ -10,12 +10,21 @@ struct RegionInfo {
-    float spacing;
+    float dutyCycle;    // modified by getEffectiveDutyCycle
+    float spacing;      // gaps between radio channels
+    float padding;      // padding at each side of the "operating channel"
```
@caveman99 this is the method we're using to centralise the frequency slots within the allowed bands.

NomDeTom commented on src/mesh/MeshRadio.h:
```diff
@@ -10,12 +10,21 @@ struct RegionInfo {
     bool audioPermitted;
     bool freqSwitching;
     bool wideLora;
+    bool licensedOnly;        // Only allow in HAM mode
```
this is unused for now, but lays the ground work for the hambands

NomDeTom reviewed (commented):
Some notes

NomDeTom committed:
committing changes to protobufs

NomDeTom committed:
Merge remote-tracking branch 'origin/develop' into New_presets

NomDeTom committed:
Update protobuf submodule to latest commit

NomDeTom committed:
Merge branch 'develop' into New_presets
